### PR TITLE
Send errors in stream instead of a grpc error from streaming rpcs when transaction or reserved connection is acquired

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

The old streaming execution that started a transaction or a reserved connection does not propagate the execution error message back. The change done in https://github.com/vitessio/vitess/pull/11592 was to fix the error propagation. The change resulted in missing propagating the open transaction or reserved connection information back in case of failure.

This PR handles those cases and adds additional tests/asserts to ensure it fixes those issues.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/11674

## Checklist

-   [X] "Backport me!" label has been added if this change should be backported 15.0, 14.0, 13.0
-   [X] Tests were added or are not required

### CODE REVIEW

**Review by dbussink — APPROVED** [review]
This was a regression introduced in the bug fix in #11592 where a different bug was fixed. The changes here make sure but the original bug from #11592 doesn't happen anymore, but also the regression is fixed and we have test coverage now for it as well.

**Review by GuptaManan100 — APPROVED** [review]

**Review by shlomi-noach — APPROVED** [review]

**Comment by deepthi** [comment]
@harshit-gangal do we have an issue? If not can you please create one that describes the bug and link it in?

**Comment by harshit-gangal** [comment]
> @harshit-gangal do we have an issue? If not can you please create one that describes the bug and link it in?

I have added the issue. Would still need ACK to merge this PR.

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
Still needs code owner approval from @systay 
Should we add @GuptaManan100 as another code owner for the grpc packages?

**Comment by GuptaManan100** [comment]
I don't mind, if that will help.

**Review by systay — APPROVED** [review]

**Pull request merged**
